### ISSUE TITLE
Get the HAR without enabling the network tab

### ISSUE THREAD

**Issue by soulgalore** [devtools-html/har-export-trigger#8]
Hi @janodvarko 
what's your feeling about bug [1436665](https://bugzilla.mozilla.org/show_bug.cgi?id=1436665) (that's the one that would enable getting the HAR without enabling the network tab right)? 

From the outside, it looks like a lot of work, so I'm wondering if it's possible to say in which version of Firefox you think that would land (without promising anything)?

Or is there a way through configuration to automate so that when you actually open Firefox from Selenium, the network tab is open and activated?

Best
Peter

**Comment by janodvarko** [comment]
> I'm wondering if it's possible to say in which version of Firefox you think that would land (without
> promising anything)?

I am shooting for Firefox 61

> Or is there a way through configuration to automate so that when you actually open Firefox from
> Selenium, the network tab is open and activated?

You can try:
1) Run Firefox with `-devtools` option on the command line to make sure that the DevTools toolbox is opened automatically
2) Set the following preference: `devtools.toolbox.selectedTool` to "netmonitor" to make sure the Net panel is selected automatically

Honza

**Comment by soulgalore** [comment]
Ok, cool thanks!

**Comment by soulgalore** [comment]
Yes I can confirm that -`devtools`  and `devtools.toolbox.selectedTool` works. I also changed the size with `devtools.toolbox.footer.height`. I'll try this out see if it works until 61 (or when it's fixed). 

Thanks again

**Comment by xmhubj** [comment]
@soulgalore I still cannot get HAR by Selenium...even with argument `-devtools` and preference `devtools.toolbox.selectedTool`.   HAR export trigger addon is set(installed) via Firefox profile, also `-devtools` and `devtools.toolbox.selectedTool` added at that time... How you make this work?

Update:  All I have to reopen the devtools panel with network tab selected and reload the web page, then I can get the HAR log.

**Comment by soulgalore** [comment]
Hi @xmhubj you can checkout our implantation:
https://github.com/sitespeedio/browsertime/tree/master/lib/firefox/webdriver

Best
Peter

**Comment by xmhubj** [comment]
@soulgalore Perter, thanks for sharing your implementation with me.  I have read your code, the difference I can tell is that you're using JavaScript bindings, while I'm using Java bindings.

```
        if (enableHar) {
            File harExportFile = new File(Configuration.FF_HET_FILE);
            
            if (harExportFile.exists()) {
                profile.setPreference("devtools.toolbox.selectedTool", "netmonitor");
                profile.setPreference("devtools.toolbox.footer.height", 250);
                profile.addExtension(harExportFile);
                options.addArguments("-devtools");
            } else {
                LOG.logp(Level.SEVERE, CLASS, METHOD, logMsgPrefix + "HAR addon file not found!");
            }
        }
```

I tried to debug the harexporttrigger addon and set a breakpoint in `har-export-trigger/src/background.js`. It looks like `devtools.js` did not connect to the background script,  this results in the value of `con.devtools` is undefined...

```
# background.js
L63      if (con.devtools && message.action) {
L64        con.devtools.postMessage(message);
L65      }
```

I'm using Firefox nightly `Mozilla Firefox 61.0a1`, which version you were using to make this work? Thank you.

**Comment by soulgalore** [comment]
Hi @xmhubj I'm using latest beta 61, but it has worked since 60 and works in latest 60 stable for me too.

When you run your test, can you see that devtools is opened?

Best
Peter

**Comment by xmhubj** [comment]
@soulgalore yes, I was able to see devtools was opened and network tab was selected.  I tested it without `driver.quit()`, therefore I can play with the Firefox instance launched by webdriver.  I hunted it down, and found the devtools.js did not setup a connection with background.js for some reason.

**Comment by xmhubj** [comment]
It's working now after I switched my Firefox version to 61 desktop version. Thanks, Peter!

**Comment by xmhubj** [comment]
Tried with Firefox 60.0esr, unfortunately, not working.... :-(.

My Firefox version:
```
Application Basics
------------------

Name: Firefox
Version: 60.0
Build ID: 20180503164101
Update Channel: esr
User Agent: Mozilla/5.0 (X11; Linux x86_64; rv:60.0) Gecko/20100101 Firefox/60.0
OS: Linux 3.13.0-85-generic
Multiprocess Windows: 1/1 (Enabled by default)
Web Content Processes: 1/4
Stylo: content = true (enabled by default), chrome = true (enabled by default)
Enterprise Policies: Inactive
Google Key: Found
Mozilla Location Service Key: Found
Safe Mode: false
```
- Network tab selected via selenium webdriver(geckodriver):

![network_tab_selected](https://user-images.githubusercontent.com/15373664/40037278-1e1cde4c-583f-11e8-9950-b381047240f8.jpg)

- No HAR log can be retrieved

![no_harlog_returned](https://user-images.githubusercontent.com/15373664/40037318-58f7ce1e-583f-11e8-9c71-38e43f974071.jpg)

- `devtools.js` from the HAR export trigger cannot be found from add-on debugger

![devtools_not_available](https://user-images.githubusercontent.com/15373664/40037366-86395334-583f-11e8-8726-a4a6b4903919.jpg)

- `devtools.js` can be seen after reopening devtools panel from add-on debugger

![reopend_devtoolsjs_available](https://user-images.githubusercontent.com/15373664/40037393-a747f90e-583f-11e8-9201-8f13240200f2.jpg)

@janodvarko any idea on what can cause the issue? Thanks.

**Comment by xmhubj** [comment]
I can manually reproduce this issue with Firefox 60.0esr:
1. Launch a firefox, the default `about:blank` is opened.
2. `shift+F5` to open devtools panel, and select `network` tab.
3. Open another browser tab and navigate to `Add-ons Manager`(`about:addons`), and install HARExportTrigger.
4. Back to the 1st tab, open a web page, say https://www.google.com
5. Select `console` tab to retrieve the HAR log.

By those steps, I cannot get the HAR log.  I took a look at https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Extending_the_developer_tools, it says:
```
The devtools page is loaded when the browser devtools are opened, and unloaded when it is closed.
```

It seems that the devtools panel opened before HAR export trigger installation, thus `devtools page` hasn't been loaded.  @soulgalore can your code work with Firefox 60.0esr?

@janodvarko As we would like to stick with Firefox ESR version(Most enterprise users are using ESR version), really want to make it work on Firefox ESR.  It is possible to let Firefox to install HAR export trigger first and then open devtools panel? It would be appreciated if you can provide some advises. Thanks!

**Comment by xmhubj** [comment]
@janodvarko ^^^

**Comment by janodvarko** [comment]
@xmhubj if you reopen DevTools toolbox (in between steps 3-4 or 4-5) does it help?

**Comment by janodvarko** [comment]
The reason why I am asking is that this obviously feels like WebExtension API issue that should be reported in bugzilla under "WebExtensions: DeveloperTools' component.

**Comment by xmhubj** [comment]
@jandovarko I did that couple days ago, it was working if installing add-on before opening devtools. Let me try again to give a firm answer.

**Comment by xmhubj** [comment]
@janodvarko Here is the result of reopen DevTools toolbox between steps 3-4 or 4-5:
The HAR log can be retrieved with **empty entries and pages**, see the screenshot below:

![reopen_devtools](https://user-images.githubusercontent.com/15373664/40162501-c3949574-59e6-11e8-8b13-87a7c9e85e1f.jpg)

I can firm that devtools.html and devtools.js in HAR Export trigger are loaded this time.

**Comment by xmhubj** [comment]
After reopening the devtools toolbox, what I have to do is **re-selecting network tab and refreshing the page** to load the web page, then I can get the network entry and page info, see the following screenshot:

![reselect_networktab](https://user-images.githubusercontent.com/15373664/40162722-62b3cc7e-59e7-11e8-833e-a1caa0394c8a.jpg)

**Comment by janodvarko** [comment]
If you also register `onRequestFinished` listener (can be just an empty function), does that help?
Honza

**Comment by xmhubj** [comment]
When and where should I register `onRequestFinished` listener?

**Comment by janodvarko** [comment]
Register it within your page as follows:

```js
HAR.addRequestListener(harEntry => {
  console.log("Request finished", request);
});
```

Before page load begins...

Honza

**Comment by xmhubj** [comment]
I think HAR object is not available before page load begins. Looks like I have to change src/content.js or lib/harapi.js, add above line into it, right?

PS: will try this later.

**Comment by janodvarko** [comment]
@xmhubj please see also this comment:
https://github.com/devtools-html/har-export-trigger/issues/13#issuecomment-389787795

Honza

**Comment by soulgalore** [comment]
I think I do not fully understand the flow and how it will work? As @xmhubj pointed out addRequestListener needs to happen in the plugin right when it is injected in the page but you will still miss the document request or because without the netmonitor open no request will be collected until getHar/addRequestListener?

1. You start the browser
2. You tell the browser to navigate to a URL
3. The plugin injects lib/harapi.js to the page and the HAR object is availible
4. Add addRequestListener so requests are collected
5. The page continuous to load
6. Get the HAR 

Am I understanding the flow correct then?

Best
Peter

**Comment by xmhubj** [comment]
I think we need to register the listener when `devtools.html\devtools.js` is loaded by toolbox.

In source file `devtools.js`, added a line to call onAddRequestListener() to register request listener.

```
/**
 * Connect to the background script and send initialization message.
 * This first message is needed to properly pair connection within
 * the background script relay logic.
 */
var port = chrome.runtime.connect({ name: "devtools" });
port.postMessage({
  tabId: chrome.devtools.inspectedWindow.tabId
});

// Register the request listener here, which will be called when `devtools.html` is loaded
onAddRequestListener();

/**
 * Listen for message sent from the content scope (they are relayed
 * by background script).
 */
function onMessage(message) {
  if (!message) {
    console.error("ERROR: message object doesn't exist!");
    return;
  }
...
...
```

Right after `devtools.js` setup a connection to `background.js`,  this can solve @soulgalore issue.  @janodvarko Do you think this is the right way to go?

My issue in [comment 13](https://github.com/devtools-html/har-export-trigger/issues/8#issuecomment-389739307), only happens on Firefox 60.0esr. Looks like Firefox 61 or later has fixed the issue.  But we still want to make this work on Firefox 60.0esr.....Just look for a workaround, to make sure devtools toolbox opened after HAR export trigger installed... With Firefox 60.0esr, `devtools.html/devtools.js` included in HAR export trigger cannot be loaded if devtools toolbox is opened before installing the add-on...

**Comment by soulgalore** [comment]
Yep you are right @xmhubj the request listener needs to be triggered when the extension init. Is that something that's ok for you @janodvarko ? Or is it a way to read configuration so it can be optional?

**Comment by janodvarko** [comment]
> Yep you are right @xmhubj the request listener needs to be triggered when the extension init. Is that something that's ok for you @janodvarko ? Or is it a way to read configuration so it can be optional?

This looks good to me. It means that HTTP data will *always* be collected if the extension is installed, but that's what (most of?) the users that install it want anyway.  Another advantage of doing this by default is that there is no additional configuration step and it will work out of the box.  So, let's wait and see if there is a problem and there are actually some users who want to opt out.
I filled a new issue #16 for this.

Thanks!
Honza

**Comment by janodvarko** [comment]
> My issue in comment 13, only happens on Firefox 60.0esr. Looks like Firefox 61 or later has fixed the issue. But we still want to make this work on Firefox 60.0esr.....Just look for a workaround, to make sure devtools toolbox opened after HAR export trigger installed... With Firefox 60.0esr, devtools.html/devtools.js included in HAR export trigger cannot be loaded if devtools toolbox is opened before installing the add-on...

@rpl is there a known bug in WebExtensions architecture in 60 (fixed in 61) that would cause that HARExportTrigger (DevTools extensions, using devtools.html scope) isn't loaded if devtools toolbox is opened before installing the add-on?

You can also see https://github.com/devtools-html/har-export-trigger/issues/8#issuecomment-389739307

Honza

**Comment by rpl** [comment]
> @rpl is there a known bug in WebExtensions architecture in 60 (fixed in 61) that would cause that HARExportTrigger (DevTools extensions, using devtools.html scope) isn't loaded if devtools toolbox is opened before installing the add-on?

Yes, it is [Bug 1378107 - WebExtensions devtools pages are not executed on existing (opened) developer toolboxes](https://bugzilla.mozilla.org/show_bug.cgi?id=1378107) which is an unfortunate side-effect of matching the Chrome behavior (which doesn't run the devtools page from a newly installed addon on any developer toolboxes that were already open) in the initial version of the WebExtensions devtools_page.

It was definitely an annoying behavior (especially while developing a devtools addons, because you have to close and reopen the developer toolbox every time you reload the addon), and so we fixed that behavior in the refactoring we did as part of [Bug 1394750 - State of devtools installed by addons not remembered](https://bugzilla.mozilla.org/show_bug.cgi?id=1394750), landed in Firefox 61, and so on Firefox >= 61 the devtools page for a newly installed addon is also going to run on the toolboxes that were already opened.

**Comment by xmhubj** [comment]
@janodvarko @rpl Thank you both! Now I'm clear.
